pkg/compare/container_test.go

```diff
@@ -0,0 +1,326 @@
+	}
+
+	for _, test := range tests {
+		// Customize behavior based on test case
+		lookPath = func(cmd string) (string, error) {
+			if (test.available == "podman" && cmd == "podman") || (test.available == "docker" || test.available == "docker-sudo") && cmd == "docker" {
+				return "/usr/bin/" + cmd, nil
+			}
+			return "", errors.New("not found")
+		}
+
+		execCommand = func(command string, args ...string) *exec.Cmd {
+			cs := []string{"-test.run=TestHelperProcess", "--", command}
+			cs = append(cs, args...)
+			cmd := exec.Command(os.Args[0], cs...)
+			cmd.Env = []string{"GO_WANT_HELPER_PROCESS=1"}
+			if test.available == "docker-sudo" {
+				cmd.Env = append(cmd.Env, "MOCK_DOCKER_FAIL=1") // Simulate `docker images` failing
+			}
+			return cmd
+		}
+
+		eng, err := newEngine()
+
+		if test.expectError {
+			if err == nil {
+				t.Errorf("Expected error but got nil")
+			}
+		} else {
+			if err != nil {
+				t.Errorf("Unexpected error: %v", err)
+			}
+			if eng.name != test.expectedName {
+				t.Errorf("Expected engine name %q, got %q", test.expectedName, eng.name)
+			}
+			if eng.requiresSudo != test.expectedSudo {
+				t.Errorf("Expected requiresSudo %v, got %v", test.expectedSudo, eng.requiresSudo)
+			}
+		}
+	}
+}
+
+func TestHelperProcess(t *testing.T) {
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
+		return
+	}
+
+	if os.Getenv("MOCK_DOCKER_FAIL") == "1" {
+		os.Exit(1)
+	}
+	fmt.Fprintf(os.Stdout, dockerRunResult)
+	os.Exit(0)
+}
+
+func TestPullAndRunContainer(t *testing.T) {
+	// Override execCommand with a fake function
+	execCommand = fakeExecCommand
+	defer func() { execCommand = exec.Command }()
+
+	tests := []struct {
+		engine      engine
+		image       string
+		expectError bool
+	}{
+		{engine{"podman", false, "", ""}, "hello-world", false},
+		{engine{"docker", true, "", ""}, "hello-world", false},
+		{engine{"docker", true, "", ""}, "invalid-image", true},
+	}
+
+	for _, test := range tests {
+		err := test.engine.pullAndRunContainer(test.image)
+		if test.expectError {
+			if err == nil {
+				t.Errorf("Expected error but got nil")
+			}
+		} else {
+			if err != nil {
+				t.Errorf("Expected nil error, got %#v", err)
+			}
+			if test.engine.containerID != dockerRunResult {
+				t.Errorf("Expected container ID %q, got %q", dockerRunResult, test.engine.containerID)
+			}
+		}
+	}
+}
+
+func TestExtractReferences(t *testing.T) {
+	// Override execCommand with a fake function
+	execCommand = fakeExecCommand
+	defer func() { execCommand = exec.Command }()
+
+	tests := []struct {
+		engine         engine
+		pathToMetadata string
+		dname          string
+		expectError    bool
+	}{
+		{engine{"podman", false, "container123", ""}, "/etc/configs", "/tmp/extracted", false},
+		{engine{"docker", true, "container456", ""}, "/var/lib/data", "/tmp/extracted", false},
+		{engine{"docker", true, "container456", ""}, "/invalid_path", "/tmp/extracted", true},
```

Comment:
Other test case ideas:
- What if tmpdir does not exist (Or is that relevant at this level?)